src/hazelcore/dynamics/EvaluatorStep.re

```diff
@@ -0,0 +1,490 @@
+  | Inj(_, _, d1) => is_boxedval(d1)
+  | Cons(d1, d2) => is_boxedval(d1) && is_boxedval(d2)
+  | Cast(d1, ty, ty') =>
+    switch (
+      EvaluatorCommon.ground_cases_of(ty),
+      EvaluatorCommon.ground_cases_of(ty'),
+    ) {
+    | (Ground, Hole)
+    | (Ground, NotGroundOrHole(_))
+    | (NotGroundOrHole(_), Ground) => is_boxedval(d1)
+    | (NotGroundOrHole(_), NotGroundOrHole(_)) =>
+      !HTyp.eq(ty, ty') && is_boxedval(d1)
+    | (Hole, Ground) =>
+      switch (d1) {
+      | Cast(_, _, Hole) => false
+      | _ => is_boxedval(d1)
+      }
+    | _ => false
+    }
+  | _ => false
+  };
+
+let rec is_final = (d: DHExp.t): bool => is_boxedval(d) || is_indet(d)
+and is_indet = (d: DHExp.t): bool =>
+  switch (d) {
+  | Ap(Cast(_, Arrow(_, _), Arrow(_, _)), _) => false
+  | FreeVar(_)
+  | Keyword(_)
+  | InvalidText(_)
+  | EmptyHole(_, _, _) => true
+  | NonEmptyHole(_, _, _, _, d1) => is_final(d1)
+  | InvalidOperation(d1, _) =>
+    is_final(d1)
+    || (
+      switch (d1) {
+      | BinIntOp(Divide, _, IntLit(0)) => true
+      | _ => false
+      }
+    )
+  | Cast(d1, ty, ty') =>
+    switch (
+      EvaluatorCommon.ground_cases_of(ty),
+      EvaluatorCommon.ground_cases_of(ty'),
+    ) {
+    | (Ground, Hole)
+    | (Ground, NotGroundOrHole(_))
+    | (NotGroundOrHole(_), Ground) => is_indet(d1)
+    | (NotGroundOrHole(_), NotGroundOrHole(_)) =>
+      !HTyp.eq(ty, ty') && is_indet(d1)
+    | (Hole, Ground) =>
+      switch (d1) {
+      | Cast(_, _, Hole) => false
+      | _ => is_indet(d1)
+      }
+    | _ => false
+    }
+  | FailedCast(d1, _, _) => is_final(d1)
+  // && EvaluatorCommon.ground_cases_of(ty1) == Ground
+  // && EvaluatorCommon.ground_cases_of(ty2) == Ground
+  // && !HTyp.eq(ty1, ty2) // why we need this?
+  | Let(dp, d1, _) =>
+    switch (is_final(d1), Elaborator_Exp.matches(dp, d1)) {
+    | (true, Matches(_)) => true
+    | _ => false
+    }
+  | Ap(d1, d2) =>
+    is_final(d1)
+    && (
+      switch (d1) {
+      | Lam(dp, _, _) =>
+        is_final(d2)
+        && (
+          switch (Elaborator_Exp.matches(dp, d2)) {
+          | DoesNotMatch
+          | Indet => true
+          | _ => false
+          }
+        )
+      | Cast(_, Arrow(_, _), Arrow(_, _)) => false
+      | _ => is_indet(d1) && is_final(d2)
+      }
+    )
+  | Inj(_, _, d1) => is_indet(d1)
+  | Pair(d1, d2) => is_indet(d1) && is_final(d2) || is_indet(d2)
+  | Cons(d1, d2) => is_indet(d1) && is_final(d2) || is_indet(d2)
+  | BinIntOp(_, d1, d2) => is_indet(d1) && is_final(d2) || is_indet(d2)
+  | BinBoolOp(_, d1, d2) => is_indet(d1) && is_final(d2) || is_indet(d2)
+  | BinFloatOp(_, d1, d2) => is_indet(d1) && is_final(d2) || is_indet(d2)
+  | InconsistentBranches(_, _, _, Case(d1, rules, n))
+  | ConsistentCase(Case(d1, rules, n)) =>
+    is_final(d1)
+    && (
+      switch (List.nth_opt(rules, n)) {
+      | None => true
+      | Some(Rule(dp, _)) => Elaborator_Exp.matches(dp, d1) == Indet
```

Comment:
you have to keep going down the rules and find one that is Indet if the current one does not match

Reply:
or if they all do not match, it is indet due to match failure

Reply:
I regard each matches as a step. So, if it matches, it will be put into evaluator and in the evaluator, the index will be added one.

Reply:
see comment on Slack -- I think it makes sense to fold is_indet and is_boxedval into the step function to avoid issues like this